src/plan.rs

```diff
@@ -0,0 +1,504 @@
+    fn lookup_hash256(&self, _: &Pk::Hash256) -> bool {
+        false
+    }
+
+    /// Given a RIPEMD160 hash, look up its preimage
+    fn lookup_ripemd160(&self, _: &Pk::Ripemd160) -> bool {
+        false
+    }
+
+    /// Given a HASH160 hash, look up its preimage
+    fn lookup_hash160(&self, _: &Pk::Hash160) -> bool {
+        false
+    }
+
+    /// Assert whether a relative locktime is satisfied
+    fn check_older(&self, _: Sequence) -> bool {
+        false
+    }
+
+    /// Assert whether an absolute locktime is satisfied
+    fn check_after(&self, _: LockTime) -> bool {
+        false
+    }
+}
+
+/// Wrapper around [`Assets`] that logs every query and value returned
+#[cfg(feature = "std")]
+pub struct LoggerAssetProvider(Assets);
+
+#[cfg(feature = "std")]
+macro_rules! impl_log_method {
+    ( $name:ident, $( <$ctx:ident: ScriptContext > )? $( $arg:ident : $ty:ty, )* -> $ret_ty:ty ) => {
+        fn $name $( <$ctx: ScriptContext> )? ( &self, $( $arg:$ty ),* ) -> $ret_ty {
+            let ret = (self.0).$name $( ::<$ctx> )*( $( $arg ),* );
+            dbg!(stringify!( $name ), ( $( $arg ),* ), &ret);
+
+            ret
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl AssetProvider<DefiniteDescriptorKey> for LoggerAssetProvider {
+    impl_log_method!(lookup_ecdsa_sig, pk: &DefiniteDescriptorKey, -> bool);
+    impl_log_method!(lookup_tap_key_spend_sig, pk: &DefiniteDescriptorKey, -> bool);
+    impl_log_method!(lookup_tap_leaf_script_sig, pk: &DefiniteDescriptorKey, leaf_hash: &TapLeafHash, -> bool);
+    impl_log_method!(lookup_raw_pkh_pk, <Ctx: ScriptContext> hash: &hash160::Hash, -> Option<usize>);
+    impl_log_method!(lookup_raw_pkh_x_only_pk, hash: &hash160::Hash, -> bool);
+    impl_log_method!(lookup_raw_pkh_ecdsa_sig, <Ctx: ScriptContext> hash: &hash160::Hash, -> Option<usize>);
+    impl_log_method!(lookup_raw_pkh_tap_leaf_script_sig, hash: &(hash160::Hash, TapLeafHash), -> bool);
+    impl_log_method!(lookup_sha256, hash: &sha256::Hash, -> bool);
+    impl_log_method!(lookup_hash256, hash: &hash256::Hash, -> bool);
+    impl_log_method!(lookup_ripemd160, hash: &ripemd160::Hash, -> bool);
+    impl_log_method!(lookup_hash160, hash: &hash160::Hash, -> bool);
+    impl_log_method!(check_older, s: Sequence, -> bool);
+    impl_log_method!(check_after, t: LockTime, -> bool);
+}
+
+impl<T, Pk> AssetProvider<Pk> for T
+where
+    T: Satisfier<Pk>,
+    Pk: MiniscriptKey + ToPublicKey,
+{
+    fn lookup_ecdsa_sig(&self, pk: &Pk) -> bool {
+        Satisfier::lookup_ecdsa_sig(self, pk).is_some()
+    }
+
+    fn lookup_tap_key_spend_sig(&self, _: &Pk) -> bool {
+        Satisfier::lookup_tap_key_spend_sig(self).is_some()
+    }
+
+    fn lookup_tap_leaf_script_sig(&self, pk: &Pk, leaf_hash: &TapLeafHash) -> bool {
+        Satisfier::lookup_tap_leaf_script_sig(self, pk, leaf_hash).is_some()
+    }
+
+    fn lookup_raw_pkh_pk<Ctx: ScriptContext>(&self, hash: &hash160::Hash) -> Option<usize> {
```

Comment:
I have observed this at multiple places that there are two cases.
1. We have PK and we are trying to check for other assests and marking it as bool.
2. We have PKH and we are using something similar to this `lookup_raw_pkh` where we are returning the lenght of the key if found. But why lenght ? Significance ?

Also, What this the context behind <Ctx: ScriptContext> ? What it tells us to find ?